lib/Sema/ConstraintSystem.cpp

```diff
@@ -1175,7 +1175,8 @@ ConstraintSystem::getTypeOfMemberReference(
          func->hasArchetypeSelf())) {
       openedType = openedType->replaceCovariantResultType(
                      baseObjTy,
-                     func->getNumParameterLists());
+                     func->getNumParameterLists(),
+                     !baseObjTy->getAnyOptionalObjectType());
```

Comment:
Ok, so a thing I've figured is that - the test fails only when oops() is declared in the protocol itself, if I remove it from protocol and leave in the declaration everything works just fine, same is true for your example as well, if I add it to the protocol it fails:

```
(swift) protocol P {
            func f() -> Self?
        }
(swift)
(swift) extension P {
            func f() -> Self? { return self }
        }
(swift)
(swift) extension ImplicitlyUnwrappedOptional: P { }
Assertion failed: (!newResultType->getAnyOptionalObjectType()), function replaceCovariantResultType, file /Users/xedin/work/c++/swift-env/swift/lib/AST/Type.cpp, line 737.
0  swift                    0x000000010a3db1f6 llvm::sys::RunSignalHandlers() + 86
1  swift                    0x000000010a3dc8a9 SignalHandler(int) + 361
2  libsystem_platform.dylib 0x00007fffde143bba _sigtramp + 26
```

So I withdraw my previous comment about `ErrorType` and will dig deeper into this...